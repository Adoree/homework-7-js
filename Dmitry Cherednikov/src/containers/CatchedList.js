import React, { Component } from 'react';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';

import List from '../components/List';
import FetchError from '../components/FetchError';
import { getCatchedPokemons, getError, getIsFetching, getCatchedPage, getFetchedAllCatched } from '../reducers';
import { fetchCatchedPokemons } from '../actions';

export class CatchedList extends Component {
  static propTypes = {
    list: PropTypes.array.isRequired,
    catchedPage: PropTypes.number.isRequired,
    fetchCatchedPokemons: PropTypes.func.isRequired,
    fetchedAllCatched: PropTypes.bool.isRequired,
    isFetching: PropTypes.bool.isRequired,
    errorMessage: PropTypes.string,
  };

  static defaultProps = {
    fetchCatchedPokemons: () => {},
  };

  componentDidMount() {
    if (this.props.catchedPage === 1 && !this.props.fetchedAllCatched) {
      this.fetchPokemons();
    }
  }

  fetchPokemons = () => {
    this.props.fetchCatchedPokemons();
  };

  render() {
    const { list, errorMessage, isFetching, fetchedAllCatched } = this.props;

    if (errorMessage && !list.length) {
      return (
        <FetchError
          message={errorMessage}
          onRetry={this.fetchPokemons}
        />
      )
    }

    if (!list.length && !isFetching) {
      return (
        <p>
          There are no catched pokemons
        </p>
      )
    }

    return (
      <List
        catched
        list={list}
        onClick={this.fetchPokemons}
        isFetching={isFetching}
        isFetchedAll={fetchedAllCatched}
      />
    )
  }
}

const mapStateToProps = (state) => ({
  list: getCatchedPokemons(state),
  errorMessage: getError(state),
  fetchedAllCatched: getFetchedAllCatched(state),
  isFetching: getIsFetching(state),
  catchedPage: getCatchedPage(state),
});

const mapDispatchToProps = (dispatch) => ({
  fetchCatchedPokemons: bindActionCreators(fetchCatchedPokemons, dispatch)
});

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(CatchedList);