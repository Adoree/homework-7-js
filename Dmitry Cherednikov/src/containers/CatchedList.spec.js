import React, { Component } from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { CatchedList } from './CatchedList';

const data = [
  {name: "bulbasaur", id: 1, date: "15.07.2018, 16:28:48"},
  {name: "ivysaur", id: 2, date: "15.07.2018, 15:46:37"},
  {name: "venusaur", id: 3, date: "15.07.2018, 16:28:48"},
  {name: "charmander", id: 4, date: "15.07.2018, 16:28:49"},
  {name: "charmeleon", id: 5, date: "15.07.2018, 16:28:50"},
  {name: "charizard", id: 6, date: "15.07.2018, 16:28:49"},
  {name: "squirtle", id: 7, date: "15.07.2018, 17:56:21"},
  {name: "wartortle", id: 8, date: "15.07.2018, 17:56:20"},
  {name: "blastoise", id: 9, date: "15.07.2018, 17:56:23"},
  {name: "butterfree", id: 12, date: "15.07.2018, 17:57:45"},
];

const setup = () => {
  const props = {
    list: data,
    error: null,
    fetchedAllCatched: false,
    isFetching: false,
    catchedPage: 1,
    fetchCatchedPokemons: jest.fn(),
  };

  const output = shallow((
    <CatchedList { ...props } />
  ));

  return {
    output,
    props,
  }
};

describe('<CatchedList />', () => {
  it('should render correctly', () => {
    const { output } = setup();

    expect(toJson(output)).toMatchSnapshot();
  });

  it('should render FetchError if there is an error and we have no pokemons', () => {
    const { output } = setup();
    const errorMessage = 'Something went wrong';

    expect(output.find('FetchError')).toHaveLength(0);
    output.setProps({ list: [], errorMessage });
    expect(output.find('FetchError')).toHaveLength(1);
  });

  it('should render special message if list is empty', () => {
    const { output } = setup();
    const message = 'There are no catched pokemons';

    expect(output.find('p')).toHaveLength(0);

    output.setProps({ list: [] });

    expect(output.find('p')).toHaveLength(1);
    expect(output.find('p').text()).toEqual(message);
  });
});