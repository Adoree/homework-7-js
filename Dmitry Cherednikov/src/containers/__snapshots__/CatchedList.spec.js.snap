// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CatchedList /> should render correctly 1`] = `
<List
  catched={true}
  isFetchedAll={false}
  isFetching={false}
  list={
    Array [
      Object {
        "date": "15.07.2018, 16:28:48",
        "id": 1,
        "name": "bulbasaur",
      },
      Object {
        "date": "15.07.2018, 15:46:37",
        "id": 2,
        "name": "ivysaur",
      },
      Object {
        "date": "15.07.2018, 16:28:48",
        "id": 3,
        "name": "venusaur",
      },
      Object {
        "date": "15.07.2018, 16:28:49",
        "id": 4,
        "name": "charmander",
      },
      Object {
        "date": "15.07.2018, 16:28:50",
        "id": 5,
        "name": "charmeleon",
      },
      Object {
        "date": "15.07.2018, 16:28:49",
        "id": 6,
        "name": "charizard",
      },
      Object {
        "date": "15.07.2018, 17:56:21",
        "id": 7,
        "name": "squirtle",
      },
      Object {
        "date": "15.07.2018, 17:56:20",
        "id": 8,
        "name": "wartortle",
      },
      Object {
        "date": "15.07.2018, 17:56:23",
        "id": 9,
        "name": "blastoise",
      },
      Object {
        "date": "15.07.2018, 17:57:45",
        "id": 12,
        "name": "butterfree",
      },
    ]
  }
  onClick={[Function]}
/>
`;
